working/0723-static-extensions/feature-specification-variant1.md

```diff
@@ -0,0 +1,467 @@
+`Comparator<X>`, but a static extension provides an extra constructor that
+does not require the `Comparator<X>` argument. This extra constructor would
+have a constraint on the actual type argument, namely that it is an `X` such
+that `X extends Comparable<X>`.
+
+```dart
+class SortedList<X> {
+  final Comparator<X> _comparator;
+  SortedList(Comparator<X> this._comparator);
+  // ... lots of stuff that doesn't matter here ...
+}
+
+static extension<X extends Comparable<X>> on SortedList<X> {
+  SortedList.ofComparable(): this((X a, X b) => a.compareTo(b));
+}
+```
+
+A static extension with type parameters can be specialized for specific
+values of specific type arguments by specifying actual type arguments of
+the on-class to be types that depend on each other, or types with no type
+variables:
+
+```dart
+static extension E4<X> on Map<X, List<X>> {
+  factory Map.listValue(X x) => {x: [x]};
+}
+
+var int2intList = Map.listValue(1); // Inferred as `Map<int, List<int>>`.
+// `Map<int, double>.listValue(...)` is an error.
+
+static extension E6<Y> on Map<String, Y> {
+  factory Map.fromString(Y y) => {y.toString(): y};
+}
+
+var string2bool = Map.fromString(true); // Inferred as `Map<String, bool>`.
+Map<String, List<bool>> string2listOfBool = Map.fromString([]);
+```
+
+## Specification
+
+### Syntax
+
+The grammar is modified as follows:
+
+```ebnf
+<topLevelDefinition> ::= // Add a new alternative.
+    <classDeclaration> |
+    <mixinDeclaration> |
+    <extensionDeclaration> |
+    <staticExtensionDeclaration> | // New alternative.
+    <enumType> |
+    ...;
+
+<staticExtensionDeclaration> ::= // New rule.
+    'static' 'extension' <identifier>? <typeParameters>? 'on' <type>
+    '{' (<metadata> <staticExtensionMemberDeclaration>)* '}';
+
+<staticExtensionMemberDeclaration> ::= // New rule.
+    'static' <staticExtensionMethodSignature> <functionBody> |
+    <staticExtensionConstructor> |
+    'static' <staticExtensionVariableDeclaration> ';';
+
+<staticExtensionMethodSignature> ::= // New rule.
+    <functionSignature> |
+    <getterSignature> |
+    <setterSignature>;
+
+<staticExtensionConstructor> ::= // New rule.
+    <factoryConstructorSignature> <functionBody> |
+    <redirectingFactoryConstructorSignature> ';';
+
+<staticExtensionVariableDeclaration> ::= // New rule.
+    ('final' | 'const') <type>? <staticFinalDeclarationList> |
+    'late' 'final' <type>? <initializedIdentifierList> |
+    'late'? <varOrType> <initializedIdentifierList>;
+
+<constructorNameList> ::=
+    <constructorName> (',' <constructorName>)*;
+```
+
+In a static extension of the form `static extension E on C {...}` where `C`
+is an identifier or an identifier with an import prefix, we say that the
+on-class of the static extension is `C`.
+
+If `C` denotes a non-generic class, mixin, mixin class, or extension
+type then we say that the _constructor return type_ of the static extension
+is `C`.
+
+If `C` denotes a generic class then `E` is treated as
```

Comment:
Above you say "non-generic class, mixin, mixin class, or extension type", here you say "a generic class".   Is the difference deliberate?